File: ui/src/lib/wallets.ts
import { authedGetJson, authedSendJson } from "./auth";

export type WalletFamily = "evm" | "solana" | "bitcoin";
export type WalletSource = "phantom" | "metamask" | "ledger" | "manual";

export interface WalletAsset {
  symbol: string;
  amount: string;
  kind?: string;
}

export interface WalletBalance {
  assets: WalletAsset[];
  coverage: string;
}

export interface WalletAccount {
  id: string;
  address_family: WalletFamily;
  address: string;
  source: WalletSource;
  label: string;
  discovered_by: "manual" | "browser_extension";
  balance: WalletBalance | null;
  refreshed_at: string | null;
  refresh_error: string | null;
  created_at: string | null;
}

export interface WalletsResponse {
  accounts: WalletAccount[];
  security: {
    read_only: boolean;
    stores_private_keys: boolean;
    stores_seed_phrases: boolean;
  };
}

export interface AddWalletInput {
  address_family: WalletFamily;
  address: string;
  source: WalletSource;
  label: string;
  discovered_by?: "manual" | "browser_extension";
}

export const getWallets = (): Promise<WalletsResponse> =>
  authedGetJson<WalletsResponse>("/wallets");

export const addWallet = (input: AddWalletInput): Promise<WalletAccount> =>
  authedSendJson<WalletAccount>("POST", "/wallets", input);

export const renameWallet = (id: string, label: string): Promise<WalletAccount> =>
  authedSendJson<WalletAccount>("PATCH", `/wallets/${id}`, { label });

export const removeWallet = (id: string): Promise<{ removed: boolean }> =>
  authedSendJson<{ removed: boolean }>("DELETE", `/wallets/${id}`);

export const refreshWallet = (id: string): Promise<WalletAccount> =>
  authedSendJson<WalletAccount>("POST", `/wallets/${id}/refresh`);

export const refreshWallets = (): Promise<{ accounts: WalletAccount[] }> =>
  authedSendJson<{ accounts: WalletAccount[] }>("POST", "/wallets/refresh");

export function shortAddress(address: string): string {
  return address.length <= 16 ? address : `${address.slice(0, 7)}…${address.slice(-6)}`;
}

export function sourceName(source: WalletSource): string {
  return { phantom: "Phantom", metamask: "MetaMask", ledger: "Ledger", manual: "Address" }[source];
}

export function familyName(family: WalletFamily): string {
  return { evm: "Ethereum", solana: "Solana", bitcoin: "Bitcoin" }[family];
}

File: src/omni/api/wallets.py
"""Authenticated, read-only consumer wallet accounts."""

from __future__ import annotations

import json
from datetime import datetime
from uuid import UUID

from neutron import App, Router
from neutron.error import bad_request, conflict, not_found, unauthorized
from pydantic import BaseModel
from starlette.requests import Request

from omni import wallets
from omni.auth import resolve_audience_from_request


def _require_user(request: Request) -> UUID:
    user = resolve_audience_from_request(request)
    if user is None:
        raise unauthorized("Authentication required")
    return user


def _iso(value: datetime | None) -> str | None:
    return value.isoformat() if value is not None else None


def _payload(row) -> dict:
    balance = row["balance"]
    if isinstance(balance, str):
        balance = json.loads(balance)
    return {
        "id": str(row["id"]),
        "address_family": row["address_family"],
        "address": row["address"],
        "source": row["source"],
        "label": row["label"],
        "discovered_by": row["discovered_by"],
        "balance": balance,
        "refreshed_at": _iso(row["refreshed_at"]),
        "refresh_error": row["refresh_error"],
        "created_at": _iso(row["created_at"]),
    }


class AddWalletIn(BaseModel):
    address_family: str
    address: str
    source: str
    label: str
    discovered_by: str = "manual"


class RenameWalletIn(BaseModel):
    label: str


def build_router(app: App) -> Router:
    router = Router()

    @router.get("/wallets")
    async def list_wallets(request: Request) -> dict:
        user = _require_user(request)
        rows = await wallets.accounts_for_user(app.db.pool, user_id=user)
        return {
            "accounts": [_payload(row) for row in rows],
            "security": {
                "read_only": True,
                "stores_private_keys": False,
                "stores_seed_phrases": False,
            },
        }

    @router.post("/wallets")
    async def add_wallet(body: AddWalletIn, request: Request) -> dict:
        user = _require_user(request)
        try:
            row = await wallets.add_account(
                app.db.pool,
                user_id=user,
                address_family=body.address_family,
                address=body.address,
                source=body.source,
                label=body.label,
                discovered_by=body.discovered_by,
            )
        except wallets.DuplicateWallet as exc:
            raise conflict(str(exc)) from exc
        except ValueError as exc:
            raise bad_request(str(exc)) from exc
        return _payload(row)

    @router.patch("/wallets/{account_id}")
    async def rename_wallet(
        account_id: UUID, body: RenameWalletIn, request: Request,
    ) -> dict:
        user = _require_user(request)
        try:
            row = await wallets.rename_account(
                app.db.pool, user_id=user, account_id=account_id, label=body.label,
            )
        except ValueError as exc:
            raise bad_request(str(exc)) from exc
        if row is None:
            raise not_found("Wallet account not found")
        return _payload(row)

    @router.delete("/wallets/{account_id}")
    async def remove_wallet(account_id: UUID, request: Request) -> dict:
        user = _require_user(request)
        removed = await wallets.remove_account(
            app.db.pool, user_id=user, account_id=account_id,
        )
        if not removed:
            raise not_found("Wallet account not found")
        return {"removed": True}

    @router.post("/wallets/refresh")
    async def refresh_wallets(request: Request) -> dict:
        user = _require_user(request)
        current = await wallets.accounts_for_user(app.db.pool, user_id=user)
        refreshed = []
        for account in current[:20]:
            row = await wallets.refresh_account(
                app.db.pool, user_id=user, account_id=account["id"],
            )
            if row is not None:
                refreshed.append(_payload(row))
        return {"accounts": refreshed}

    @router.post("/wallets/{account_id}/refresh")
    async def refresh_wallet(account_id: UUID, request: Request) -> dict:
        user = _require_user(request)
        row = await wallets.refresh_account(
            app.db.pool, user_id=user, account_id=account_id,
        )
        if row is None:
            raise not_found("Wallet account not found")
        return _payload(row)

    return router


__all__ = ["build_router"]
